spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/MiniBatch.scala

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.dataset
+
+import com.intel.analytics.bigdl.nn.abstractnn.Activity
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.utils.T
+import org.apache.log4j.Logger
+
+trait MiniBatch[T] {
+  def toTensorMiniBatch[T](): TensorMiniBatch[T] = {
+    this.asInstanceOf[TensorMiniBatch[T]]
+  }
+
+  def toArrayTensorMiniBatch[T](): ArrayTensorMiniBatch[T] = {
+    this.asInstanceOf[ArrayTensorMiniBatch[T]]
+  }
+
+  /**
+   * get size of this MiniBatch
+   * @return size
+   */
+  def size(): Int
+
+  /**
+   * Check if data's batchSize match labels' batchSize
+   */
+  def selfCheck(): Unit
+
+  /**
+   * Check if batchSize meet the best performance.
+   * @param modelNumber number of models
+   */
+  def performanceCheck(modelNumber: Int): Unit = {
+    selfCheck()
+    require((size() >= modelNumber) &&
+      (size() % modelNumber == 0), "performanceCheck failed: total batch size " +
+      s"${size()} should be divided by total core number ${modelNumber}")
+    if (size() < modelNumber * 2) {
+      MiniBatch.logger.warn("Warning: for better training speed, " +
+        "total batch size is recommended to be at least two times of core number" +
+        s"${modelNumber}, please tune your batch size accordingly")
+    }
+  }
+
+  /**
+   * Copy this [[MiniBatch]] to an Array of (Activity, Activity)
+   * @param dst destination Array
+   */
+  def copyTo[T](dst: Array[(Activity, Activity)]): Unit = {
+    val stackSize = size() / dst.length
+    var b = 0
+    while (b < dst.length) {
+      dst(b) = toActivity(b * stackSize + 1, stackSize)
+      b += 1
+    }
+  }
+
+  /**
+   * Slice [[MiniBatch]] to (input, target) with offset and length
+   * @param offset offset, counted from 1
+   * @param length length
+   * @tparam T numeric type
+   * @return (input, target)
+   */
+  def toActivity[T](offset: Int, length: Int): (Activity, Activity)
```

Comment:
rename to slice and allow to use buffer here.
def slice(buffer: (Activity, Activity), offset: Int, length: Int)